sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -1935,8 +1935,20 @@ pi_result piMemBufferCreate(pi_context Context, pi_mem_flags Flags, size_t Size,
                              &Ptr));
   }
 
-  if ((Flags & PI_MEM_FLAGS_HOST_PTR_USE) != 0 ||
-      (Flags & PI_MEM_FLAGS_HOST_PTR_COPY) != 0) {
+  // As of now, zeMemAllocHost() does not support allocation of pinned
+  // host memory. So even if PI_MEM_FLAGS_HOST_PTR_ALLOC set, it allocates
+  // pageable host memory.
+  if ((Flags & PI_MEM_FLAGS_HOST_PTR_ALLOC) != 0) {
+    ze_host_mem_alloc_desc_t ZeHostDesc = {};
+    ZeHostDesc.flags = 0;
+
+    ZE_CALL(zeMemAllocHost(Context->ZeContext, &ZeHostDesc, Size,
+                           1, // TODO: alignment
+                           &HostPtr));
+    ZE_CALL(zeCommandListAppendMemoryCopy(Context->ZeCommandListInit, Ptr,
```

Comment:
Please explain why would anyone need the buffer to be allocated in both device and host memory at the same time? And why a copy of just allocated memory (uninitialized) is wanted?

Reply:
use_pinned_host_memory extension is introduced to support those sycl BEs that can "accelerate copies between host and device by allocating pinned memory". As of now L0 don't do that so this allocation is not going to improve or help in our case. It's my understanding that when PI_MEM_FLAGS_HOST_PTR_ALLOC flag is set we can either 1) report unsupported for it (current) or 2) allocate regular host memory (assuming it wasn't done) and do a device <-> host mapping. Please let me know if you have any suggestion on how to handle it.
Uninitialized mem copy is not necessary I agree.

Reply:
> use_pinned_host_memory extension is introduced to support those sycl BEs that can "accelerate copies between host and device by allocating pinned memory"

@rbegam please point me to the extension definition

> L0 don't do that so this allocation is not going to improve or help in our case.

Performance put aside for a moment, how host memory allocation is going to be useful for the pinned memory feature?

> and do a device <-> host mapping (like cuda).

I am missing to see how this "mapping" is performed.

@romanovvlad could you clarify the circumstances at which SYCL RT calls the `piMemBufferCreate` with `PI_MEM_FLAGS_HOST_PTR_ALLOC` and what's the expectation there?

Reply:
The extension definition is here: https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/UsePinnedMemoryProperty/UsePinnedMemoryPropery.adoc

> @romanovvlad could you clarify the circumstances at which SYCL RT calls the piMemBufferCreate with PI_MEM_FLAGS_HOST_PTR_ALLOC and what's the expectation there?

When a user passes `sycl::ext::oneapi::property::buffer::use_pinned_host_memory` property to the buffer constructor.
I would expect that an implementation allocates a buffer optimized for host<->device copy if the property is passed.

Reply:
The only semantic description given so far is this:
> Some SYCL backends can accelerate copies between host and device by allocating pinned memory.

1. What should do the BE that can't accelerate?
2. How (and why) `sycl::ext::oneapi::property::buffer::use_pinned_host_memory` translates into `PI_MEM_FLAGS_HOST_PTR_ALLOC`? The later does not seem to hold the same semantics.

Reply:
1. I think BE should just ignore it, but @jbrodman had a different opinion on that.
2. I tend to agree. This is how it's done currently. Could you please tell if you think we need to introduce a new flag?

Reply:
@jbrodman please share your opinion here?
@romanovvlad , are there other cases where PI_MEM_FLAGS_HOST_PTR_ALLOC is used?

Reply:
> @romanovvlad , are there other cases where PI_MEM_FLAGS_HOST_PTR_ALLOC is used?

In the SYCL RT - no.

Reply:
there is a recent commit for buffer management https://github.com/intel/llvm/commit/2ae1bc9e32c3a2dc02add5694d593ad4027d9cc2. For the case where there is an integrated device, this change introduces the similar behavior as pinned memory. 
@romanovvlad is there any correlation between PI_MEM_FLAGS_HOST_PTR_ALLOC and integrated device/ZE_DEVICE_PROPERTY_FLAG_INTEGRATED?

Reply:
I think for integrated devices all memory is "fast" since it's shared with host. The pinned memory makes more sense for discrete devices. I'm not an expert though.

Reply:
should I just ignore the flag as @romanovvlad suggested then? @smaslov-intel 
@jbrodman ping

Reply:
I think you need to do what this asks for -- allocates buffer in host-accessible memory (https://www.khronos.org/registry/OpenCL/sdk/2.2/docs/man/html/clCreateBuffer.html). Note it is not to allocate host memory AND device memory but rather to allocate shared memory for the buffer.

Reply:
We do need a test for this. Please make sure to cover the interaction of such (pinned) allocated memory and mapping/unmapping.